### ISSUE TITLE
exiv2 0.28 breaks gimp jpeg metadata export and crashes geeqie and gwenview

### ISSUE THREAD

**Issue by rouelle** [Exiv2/exiv2#2670]
#### **Describe the bug**
I'm using Gnome on Arch Linux. Since exiv2 (0.27.6-2 -> 0.28.0-1 -> 0.28.0-2 -> 0.28.0-3) was upgraded to 0.28.0-x Gimp is not able anymore to export metadata from a jpeg file opened, modified and re-saved. A window stating "Message Image JPEG run: saving metadata failed: Failed to write image" appears and the file is saved without metadata.

geeqie and gwenview also crashed since exiv2 0.28; see https://bbs.archlinux.org/viewtopic.php?id=286369

#### **To Reproduce**
1. Open a jpeg file in Gimp, edit the file, save the file and the metadata are lost with a window saying so.
2. Strangely a jpeg from a cell phone is saved correctly with it's metadata, but a jpeg from a camera (like Sony DSC-HX80) has no metadata saved. I attached two files to test (explanatory file name).
3. Using Arch Linux, exiv2 packages 0.28.0-1, 0.28.0-2, 0.28.0-3; see https://archlinux.org/packages/extra/x86_64/exiv2/ (were upgraded 3 times).

#### **Expected behavior**
exiv2 0.27 has no problem neither exporting jpeg metadata nor crashing geeqie and gwenview

#### **Desktop (please complete the following information):**
- OS and version: Arch Linux
- Exiv2 version and source: (e.g., 0.28.0-3 from https://archlinux.org/packages/extra/x86_64)

#### **Additional context**
You may test with these two jpeg files.
![jpeg_from_camera](https://github.com/Exiv2/exiv2/assets/46628603/fedd1a8a-fec4-49a0-beb2-a893ad1194c2)
![jpeg_from_cellphone](https://github.com/Exiv2/exiv2/assets/46628603/c182c5a0-72c7-409d-8267-70c5367f91c4)

**Comment by kmilos** [comment]
Duplicate of https://github.com/Exiv2/exiv2/issues/2638

**Comment by rouelle** [comment]
The Geeqie part may be a duplicate but Gimp being unable to export metadata of certain jpeg from a camera is not...

**Comment by kevinbackhouse** [comment]
@rouelle: it would be helpful if you could provide a stack trace which shows where the problem is in exiv2.

**Comment by rouelle** [comment]
I first checked in https://wiki.archlinux.org/title/Debugging/Getting_traces to get what you asked me for.     

Running gdb gives me 612M of data in ~/.cache/debuginfod_client. I scanned the data looking for files containing the string "exiv" and in subdir of debuginfod_client that were modified today 2023-07-03 around 09:26 when I opened Gimp and tried to save the loaded 4:3 image from the Sony camera (the window saying that the metadata were not saved appeared).

I compressed the resulting debuginfo files in a zip archive (I renamed the 4 different debuginfo to distinguish them).

Hope that will be usefull...
[debuginfo-.zip](https://github.com/Exiv2/exiv2/files/11938121/debuginfo-.zip)

**Comment by kmilos** [comment]
I think this can be closed now?

**Comment by rouelle** [comment]
Well, since I opened the bug nothing has changed, I'm using exiv2 0.28.2-3 on Arch Linux. Exporting from gimp 2.10.36-5 still miss all exif metadata. So, since last year, I export my metadata with cli command:

exiftool -tagsfromfile %f.jpg -ext jpg -overwrite_original ~/Images/bla-bla-bla-dir

I'm curious to know if I'm alone with this bug? If so, I'll close it as there is a workaround even if it is not graceful...

**Comment by rouelle** [comment]
Well, since I opened the bug nothing has changed, I'm using exiv2 0.28.2-3 on Arch Linux. Exporting from gimp 2.10.36-5 still miss all exif metadata. So, since last year, I export my metadata with cli command:

exiftool -tagsfromfile %f.jpg -ext jpg -overwrite_original ~/Images/bla-bla-bla-dir

I'm curious to know if I'm alone with this bug? If so, I'll close it as there is a workaround even if it is not graceful...
Only the export of metadata doesn't work, geeqie and gwenview doesn't crash anymore.

**Comment by kmilos** [comment]
Bu the crashes are gone? We can at least update the issue tile then?

**Comment by rouelle** [comment]
Le mer 3 avril 2024 00:36:55 -0700
Miloš Komarčević ***@***.***> écrivait :
> But the crashes are gone? We can at least update the issue tile then?

Of course! But like I said I'm still curious to know if I'm alone with this bug!

-- 

Édité et transmis avec des logiciels libres sous Linux !

**Comment by gigago** [comment]
Same here, gimp discarding Sony SOOC exif metadata with error message. Consistently repeatable.

console err mesg
JPEG image-Warning: run: saving metadata failed: Failed to write image

arch, updated last minute
gimp 2.10.38
exiv2 0.28.2

**Comment by sleeksorrow** [comment]
Is there any progress or workaround? On Manjaro my workflow is seriously broken with exiv 20.28.3. Tried GIMP 2.10.38 from repos and verified with GIMP 2.99.18 from flatpak, both affected with all pictures from my Sony Alpha 6400.

**Comment by rouelle** [comment]
Still the same for me on Arch Linux. gimp discards Sony DSC-HX80 and Nikon Z5 exif metadata with error message. Consistently repeatable. And sleeksorrow is right when he says _my workflow is seriously broken_

arch, updated last minute
gimp 2.10.38
exiv2 0.28.3

**Comment by gigago** [comment]
For me, executing once

```exiftool -r -overwrite_original -Make=Sony .```

in *SONY* picture folders - replacing the value "SONY" with "Sony" in all files - before editing in gimp worked just fine. No more errors after that.

**Comment by rouelle** [comment]
You're right, I tried for a Sony DSC-HX80 and it works. I didn't check for NIKON CORPORATION. Anyways we should'nt have to change the Make tag spelling to lowercase and, IMHO, exiv2 should take care of the tag be it lower or uppercase.

**Comment by gigago** [comment]
Sure, it's a silly bug that ought to be fixed, but at least there's a workaround, so you can continue your life with just a slight nuisance.

**Comment by sleeksorrow** [comment]
> Sure, it's a silly bug that ought to be fixed, but at least there's a workaround, so you can continue your life with just a slight nuisance.

Please, with no offense to you, this comment was unnecessary and and does not match the criticality of this bug. For criticality you also have to consider risk probability in addition to risk consequence.

For now, like 4 or 5 people now have a workaround for a bug that only affects Arch and Manjaro users of GIMP. For myself I was debugging and trying and searching hours and hours until I now have a workaround I could live with.

Can you imagine what happens, as soon as Ubuntu moves to exiv2 0.28? Thousands and thousands of GIMP users will get an error popup when saving jpeg or png. Thousands wasting time searching for a solution, flooding gimp forums and later on this issue tracker. We all will probably learn a lot of new curse words by then.

We are just lucky that Ubuntu 24.10 will still stick to exiv2 0.27.6 for now. https://packages.ubuntu.com/oracular/exiv2

**Comment by gigago** [comment]
Critical or otherwise, there's many bugs of this nature in OSS. Ironing them out is a lot of work that's never finished because of regressions and usually has to be done for free. While it would be nice if everything was fixed yesterday, having a workaround available is the next best thing.

**Comment by sleeksorrow** [comment]
You obviously did not get the point I was making. A workaround is great, but you are using it's sheer existence to lower the criticality. Until GIMP is giving this hint together with the error message, this will still help only 4-5 people, the rest will flame the devs here.

**Comment by gigago** [comment]
I'm neither member of the dev team nor issue manager, so I'm prioritising precisely nothing. If this is not fixed in a timely fashion, developers may face a lot of flac down the line. That's their problem, not mine - or yours, for that matter.

**Comment by sleeksorrow** [comment]
Then your comments to belittle this issue are even more unnecessary as I thought and I have no clue what your goal even is.

**Comment by gigago** [comment]
Usually, the best way to get something fixed is to ask the right person nicely - in this case, a developer. Just saying.

**Comment by Wormnest** [comment]
We have a new GIMP issue [here](https://gitlab.gnome.org/GNOME/gimp/-/issues/12518) that seems to be the same issue. I tried testing with the two images above, but the first one downloads as a zero-length file and the second doesn't give any issues exporting in GIMP.

Changing `Make` to `Sony` in exiftool does indeed fix the exporting issue in GIMP, but I doubt that is the real problem.

I added some printf statements in exiv2 and with the zipped image in the above mentioned GIMP issue the error reported by exiv2 is `Failed to write image`. Specifically it is in `tiffcomposite_inc.c`  function `TiffDirectory::doWrite` the line where it does `enforce(sv == d, ErrorCode::kerImageWriteFailed);`.

In this case, sv is 563, while d is 1945. So it seems there is more written than expected.

Note that after opening in GIMP I directly exported without making any changes.

Comparing the exif tags of this image (in GIMP with the metadata viewer), with the one adjusted by ExifTool, I see the following differences:
- Exif.MakerNote.ByteOrder and Exif.MakerNote.Offset are missing in the ExifTool version
- Exif.Phot.MakerNote has 4 bytes less in the ExifTool version
- None of the Exif.Sony.* tags show up in the ExifTool version, though they are there when inspected with ExifTool.

**Comment by Wormnest** [comment]
> * None of the Exif.Sony.* tags show up in the ExifTool version, though they are there when inspected with ExifTool.

I wonder if that is related to [issue 2913](https://github.com/Exiv2/exiv2/issues/2913).

**Comment by Wormnest** [comment]
Some further experimenting shows that removing all `Exif.Sony.Misc3c` tags when exporting from GIMP causes saving of metadata to succeed.

**Comment by gigago** [comment]
Interesting. Seems like exiftool does more than just update the requested tag.
